### PR TITLE
Fixes #21689 - Validate content source in facet

### PR DESCRIPTION
**Pull request by akofink**
This change should check that the content source exists in the
database only if a content_source_id is given for a host.

### CODE REVIEW

File: app/models/katello/host/content_facet.rb [hunk 1]
```diff
@@ -22,6 +22,11 @@ class ContentFacet < Katello::Model
       validates :lifecycle_environment, :presence => true, :allow_blank => false
       validates :host, :presence => true, :allow_blank => false
       validates_with Validators::ContentViewEnvironmentValidator
+      validate :content_source_exists, if: :content_source_id
+
+      def content_source_exists
+        errors.add(:content_source, N_("with given ID not found")) if content_source.nil?
+      end
```
Comment (jlsherrill) [inline app/models/katello/host/content_facet.rb]: we don't create these explicit validations for every association, is something special about content_source_id?

Comment (akofink) [reply app/models/katello/host/content_facet.rb]: Nope, but how else would we "fix" the issue?

Comment (jlsherrill) [reply app/models/katello/host/content_facet.rb]: Foreman typically does lookups in the controller based on the parameter (throwing a 404 if its not found).  

This seems like a slippery slope doing it here, as we'd need to add these nice validators to every single association we have

Comment (akofink) [reply app/models/katello/host/content_facet.rb]: Ah that would work. I'll look into it Monday. Though, an initial thought: we don't really have clean controller extensions like we do for the model (aka facets) afaik. I'm not sure I agree with the slippery slope argument - we could use an [`EachValidator`](http://api.rubyonrails.org/v3.1/classes/ActiveModel/EachValidator.html). It seems like the perfect solution. Eh?

Comment (akofink) [reply app/models/katello/host/content_facet.rb]: Eh @jlsherrill? (just a ping for after holiday) :turkey:

Comment (jlsherrill) [reply app/models/katello/host/content_facet.rb]: This isn't a 'slippery slope' argument IMHO.  There is a problem that is systemic to the application and you are proposing to just solve it in a single place.  If we're not willing to put in the time and effort to fix it across the application (for example if its not worth the time and effort), i think we should punt the issue.

I had thought this was covered in foreman, but it appears not:

```
# hammer -p changeme hostgroup create --name=foobar123 --realm-id=99999
Could not create the hostgroup:
  ERROR:  insert or update on table "hostgroups" violates foreign key constraint "hostgroups_realms_id_fk"
  DETAIL:  Key (realm_id)=(99999) is not present in table "realms".

# hammer -p changeme hostgroup create --name=foobar123 --operatingsystem-id=99999
Could not create the hostgroup:
  ERROR:  insert or update on table "hostgroups" violates foreign key constraint "hostgroups_operatingsystem_id_fk"
  DETAIL:  Key (operatingsystem_id)=(99999) is not present in table "operatingsystems".
```

so really this needs to be fixed across foreman and katello.

Comment (akofink) [reply app/models/katello/host/content_facet.rb]: @jlsherrill Care to comment on the `EachValidator`? That's my proposal for a DRY way to solve this.

Comment (jlsherrill) [reply app/models/katello/host/content_facet.rb]: I might have to see a sample implementation, but i think if we could make it somewhat automatic, possibly with some override to handle non-consistent column names, that would be fine

**Comment by akofink** [comment]
I'd argue this validator should go in Foreman rather than Katello. Also, note we can pass any attributes we want to this validator on one line. I believe this is DRY enough, and I'd rather turn on the validation for each model that we want it in rather than it happening automatically (by parsing belongs_to or something). What do you think?

**Comment by jlsherrill** [comment]
As we had discussed, yeah i think this should be 'fixed' in foreman with any katello changes needed along side that fix.  @akofink we can close this now?

**Pull request closed without merging**